### ISSUE TITLE
[Enhancement] Timeout when waiting for a reply

### ISSUE THREAD

**Issue by TheCrypt0** [shogowada/json-rpc-2.0#20]
Hi,
Thank you for the library! I've been tinkering around a bit and I've successfully implemented it in my backend. (First time using node.js, yay!)

I'm now testing some stress scenario where the devices cannot reply in a reasonable amount of time or the request gets lost (maybe the MCU was busing doing something else and needs a heads up). It would be great if some kind of timeout for reply could be implemented.

Current use:
```
1. Connect WS 
2. Send RPC Request to get Device Status
3. await the reply
4. Wait 5 seconds
5. Go back to point 2 and request Dev Status
```

If the device never replies the await will never resolve and the process breaks.

Example with timeout specified:
```js
client
  .request("echo", { text: "Hello, World!" }, 2000) // <--- here's the timeout
  .then((result) => console.log(result));
```

This will also remove the request from the `idToResolveMap` once the timeout has occurred.

**Pseudo code implementation** (first time using Typescript, there is be a better way for sure for doing this, please don't be harsh 😄 )
```ts

requestAdvanced(
  request: JSONRPCRequest,
  clientParams?: ClientParams
): PromiseLike<JSONRPCResponse> {
  const promise: PromiseLike<JSONRPCResponse> = new Promise((resolve) =>
    this.idToResolveMap.set(request.id!, resolve)
  );
  // TIMEOUT STUFF - Handle the timeout 
  timeoutTimer = setTimeout(function() { 
    // Remove the request from the map
    this.idToResolveMap.delete(request.id!);
    // Throw an exception (?) in some way to let the caller know timeout has occurred
  }, 5000);
  return this.send(request, clientParams).then(
    () => promise,
    // TIMEOUT STUFF - (somewhere here) When a reply is received, clear the timeout
    clearTimeout(timeoutTimer)
    (error) => {
      this.receive(
        createJSONRPCErrorResponse(
          request.id!,
          0,
          (error && error.message) || "Failed to send a request"
        )
      );
      return promise;
    }
  );
}
```

I would love to make the PR for this feature myself but I'm still getting a grasp on how typescript works.

**Comment by shogowada** [comment]
Thank you for using the library! This looks like it has a number of use cases. Let me think about how I can implement this because `receive` method already takes 3 parameters and I want to avoid adding another parameter if possible.

**Comment by shogowada** [comment]
It is released under v0.2.18: https://github.com/shogowada/json-rpc-2.0/pull/22

Thanks for the suggestion!

**Comment by TheCrypt0** [comment]
@shogowada Thank you so much, looking forward to implement the new feature!
